src/main/java/gov/nci/ppe/controller/UserController.java

```diff
@@ -441,10 +443,43 @@ private String verifyToken(String authString) {
 		
 		List<UserEnrollmentDataDTO> userEnrollmentData = openResponseDTO.getData();
 		List<User> newUsersList = new ArrayList<>();
+		Set<Provider> providerSet = new HashSet<Provider>();
 		userEnrollmentData.forEach(patientData ->{
+			/* Verify if the provider for a particular patient is already in the system
+			 * If not insert the provider details and then associate it with them with the patient.
+			 * */
+			Optional<Provider> treatingProviderOptional = userService.findProviderByCtepId(patientData.getTreatingInvestigatorCtepId());
+			if(treatingProviderOptional.isEmpty()) {
+				Provider treatingProvider = new Provider();
+				treatingProvider.setOpenCtepID(patientData.getTreatingInvestigatorCtepId());
+				treatingProvider.setFirstName(patientData.getTreatingInvestigatorFirstName());
+				treatingProvider.setLastName(patientData.getTreatingInvestigatorLastName());
+				treatingProvider.setPhoneNumber(formatPhoneNumber(patientData.getTreatingInvestigatorPhone()));
+				treatingProvider.setEmail(patientData.getTreatingInvestigatorEmail());
+				providerSet.add((Provider)userService.insertNewProviderDetailsFromOpen(treatingProvider).get());
+			}else {
+				providerSet.add((Provider)treatingProviderOptional.get());
+			}
```

Comment:
This is a duplicate of the next section. Make it a private method and call it twice

Reply:
Also, this business logic is best placed in the Service Layer